oci/layout/oci_dest.go

```diff
@@ -184,23 +193,20 @@ func (d *ociImageDestination) PutManifest(m []byte) error {
 		Architecture: runtime.GOARCH,
 		OS:           runtime.GOOS,
 	}
-	d.index.Manifests = append(d.index.Manifests, desc)
+	d.addManifest(&desc)
 
 	return nil
 }
 
-func ensureDirectoryExists(path string) error {
-	if _, err := os.Stat(path); err != nil && os.IsNotExist(err) {
-		if err := os.MkdirAll(path, 0755); err != nil {
-			return err
+func (d *ociImageDestination) addManifest(desc *imgspecv1.Descriptor){
+	for i, manifest := range d.index.Manifests {
+		if manifest.Digest.String() == desc.Digest.String() {
```

Comment:
> I think #318 is not doing the right thing.

I was just saying that it tries to address the same issue. I did not look into the exact details.

> Use different tags, I'm not sure I understand what you mean.

Let's look at a concrete example. Say I have busybox:1.27.7 and alpine:3.1 and I want to push them  into an OCI index under _/tmp/oci_

Currently I would do something like this:

    srcRef, err := alltransports.ParseImageName("oci:/tmp/oci")

In this case the org.opencontainers.image.ref.name becomes 'latest'  - https://github.com/containers/image/blob/master/oci/layout/oci_transport.go#L97

So what options do I have? 

     srcRef, err := alltransports.ParseImageName("oci:/tmp/oci:alpine:3.1")

does not work (at least not in the current impl). I could 

    srcRef, err := alltransports.ParseImageName("oci:/tmp/oci:alpine@3.1")

In this case I create my own 'tag' by artificially combining the image name and version. Is this what I should do? This would lead to a org.opencontainers.image.ref.name of alpine@3.1. Should it not be org.opencontainers.image.ref.name = alpine and org.opencontainers.image.version = 3.1? That's what I would expect looking at the OCI specs.

What if the registry matters. Say I have a custom alpine image in myrepo - myrepo/alpine:4. How do I do this? Are you suggesting something like:

    srcRef, err := alltransports.ParseImageName("oci:/tmp/oci:myrepo@alpine@3.1")

I feel it should be possible to something along the line:

    srcRef, err := alltransports.ParseImageName("oci:/tmp/oci@myrepo/alpine:3.1")

This allows me to keep the the well known notation <repo>/<image>:<version>.

There are multiple levels to this:

* From an containers/image API point of view and the need to create src and dest references based on strings, how do I denote the concepts of base directory, image name and version
* The second aspect is the concrete implementation. Right now it assumes that there is only one image per OCI directory. Depending on the outcome of the first question, the syntax and parsing of an OCI reference needs to be changed
* Lastly there is a question of spec interpretation. I would expect that if I want to push alpine:3.1 into an OCI index, the top level entry in the index should use 'org.opencontainers.image.ref.name' as name for the image, aka 'alpine' and 'org.opencontainers.image.ref.version' as version, aka 3.1. 

Does this make sense?

 > I'd recommend taking a look at https://github.com/openSUSE/umoci's interface to see what I mean (disclosure, I wrote it).

The link does not seem to work. Can you point to the actual interface?